From bff55d04f84d7bc2db0e864757789d0698597189 Mon Sep 17 00:00:00 2001
From: bings <bings@codeaurora.org>
Date: Mon, 6 Apr 2020 10:34:16 +0800
Subject: [PATCH 6/9] sae

---
 include/net/cfg80211.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index b61f4f99902..7317bd1e08b 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -75,6 +75,9 @@ struct wiphy;
  * wireless hardware capability structures
  */
 
+/* Indicate backport support for external authentication*/
+#define CFG80211_EXTERNAL_AUTH_SUPPORT 1
+
 /**
  * enum ieee80211_channel_flags - channel flags
  *
-- 
2.49.0

